libgalois/include/katana/PropertyGraph.h

```diff
@@ -29,12 +29,81 @@ struct KATANA_EXPORT GraphTopology {
   using edges_range = StandardRange<edge_iterator>;
   using iterator = node_iterator;
 
+private:
+  LargeArray<Edge> adj_indices_;
+  LargeArray<Node> dests_;
+
+  // TODO(amber): make these private in the near future. No other class should
+  // access these directly. Instead provide access methods
+public:
   std::shared_ptr<arrow::UInt64Array> out_indices;
   std::shared_ptr<arrow::UInt32Array> out_dests;
 
-  uint64_t num_nodes() const { return out_indices ? out_indices->length() : 0; }
+private:
+  // TODO: generalize to typename T
+  static std::shared_ptr<arrow::UInt64Array> largeToArrowArray(
+      LargeArray<uint64_t>& lgArr) noexcept {
+    return std::make_shared<arrow::UInt64Array>(
+        lgArr.size(), arrow::MutableBuffer::Wrap(lgArr.data(), lgArr.size()));
+  }
+
+  static std::shared_ptr<arrow::UInt32Array> largeToArrowArray(
+      LargeArray<uint32_t>& lgArr) noexcept {
+    return std::make_shared<arrow::UInt32Array>(
+        lgArr.size(), arrow::MutableBuffer::Wrap(lgArr.data(), lgArr.size()));
+  }
+
+  GraphTopology(const GraphTopology&) = delete;
+  GraphTopology& operator=(const GraphTopology&) = delete;
+
+public:
+  GraphTopology() = default;
+
+  GraphTopology(
+      const Edge* adjIndices, size_t numNodes, const Node* dests,
+      size_t numEdges);
+
+  GraphTopology(
+      LargeArray<Edge>&& adjIndices, LargeArray<Node>&& dests) noexcept
+      : adj_indices_(std::move(adjIndices)),
+        dests_(std::move(dests)),
+        out_indices(largeToArrowArray(adj_indices_)),
+        out_dests(largeToArrowArray(dests_)) {}
+
+  static std::unique_ptr<GraphTopology> Copy(
+      const GraphTopology& that) noexcept;
 
-  uint64_t num_edges() const { return out_dests ? out_dests->length() : 0; }
+  /*
+  GraphTopology(GraphTopology&& that):
+    adj_indices_(std::move(that.adj_indices_)),
+    dests_(std::move(that.dests)),
+    out_indices(largeToArrowArray(adj_indices_)),
+    out_dests(largeToArrowArray(dests_))
+  {}
+
+  GraphTopology& operator = (GraphTopology&& that) {
+    adj_indices_ = std::move(that.adj_indices_);
+    dests_ = std::move(that.dests_);
+    out_indices = largeToArrowArray(adj_indices_);
+    out_dests = largeToArrowArray(dests_);
+    return *this;
+  }
+  */
+
+  uint64_t num_nodes() const { return adj_indices_.size(); }
+
+  uint64_t num_edges() const { return dests_.size(); }
+
+  // TODO(amber): re-enable these in the near future with better names
+  /*
+  const auto& adj_indices_arrow() const noexcept { return out_indices; }
+
+  auto& adj_indices_arrow() noexcept { return out_indices; }
+
+  const auto& dests_arrow() const noexcept { return out_dests; }
+
+  auto& dests_arrow() noexcept { return out_dests; }
+  */
```

Comment:
We may well not need these at all if we provide a clean API. If people need indexable objects we could create wrappers that expose the container concept abstractly instead of a direct arrow type.

Reply:
deleted.